parsers/CNDC.py

```diff
@@ -0,0 +1,152 @@
+SOURCE = "cndc.bo"
+
+
+def extract_xsrf_token(html):
+    """Extracts XSRF token from the source code of the generation graph page."""
+    return re.search(r'var ttoken = "([a-f0-9]+)";', html).group(1)
+
+
+def get_timestamp(query_date: datetime, hour: int) -> datetime:
+    return datetime(
+        year=query_date.year,
+        month=query_date.month,
+        day=query_date.day,
+        hour=hour,
+        tzinfo=tz_bo,
+    )
+
+
+def fetch_data(
+    session: Session | None = None, target_datetime: datetime | None = None
+) -> tuple[list[dict], datetime]:
+    if session is None:
+        session = Session()
+
+    if target_datetime is None:
+        target_datetime = datetime.now()
+    target_datetime = target_datetime.astimezone(tz_bo)
+    # Define actual and previous day (for midnight data).
+    formatted_dt = target_datetime.strftime("%Y-%m-%d")
+
+    # XSRF token for the initial request
+    xsrf_token = extract_xsrf_token(session.get(INDEX_URL).text)
+
+    resp = session.get(
+        DATA_URL.format(formatted_dt), headers={"x-csrf-token": xsrf_token}
+    )
+
+    hour_rows = json.loads(resp.text.replace("ï»¿", ""))["data"]
+    return hour_rows, target_datetime
+
+
+def parse_generation_forecast(
+    zone_key: ZoneKey, date: datetime, raw_data: list[dict], logger: Logger
+) -> TotalProductionList:
+    result = TotalProductionList(logger)
+    assert date.tzinfo == tz_bo
+    for hour_row in raw_data:
+        [hour, forecast, total, thermo, hydro, wind, solar, bagasse] = hour_row
+
+        # "hour" is one-indexed
+        timestamp = get_timestamp(query_date=date, hour=hour - 1)
+
+        result.append(
+            zoneKey=zone_key,
+            datetime=timestamp,
+            value=forecast,
+            source=SOURCE,
+            sourceType=EventSourceType.forecasted,
+        )
+
+    return result
+
+
+def parser_production_breakdown(
+    zone_key: ZoneKey, date: datetime, raw_data: list[dict], logger: Logger
+) -> ProductionBreakdownList:
+    result = ProductionBreakdownList(logger)
+    assert date.tzinfo == tz_bo
+    for hour_row in raw_data:
+        [hour, forecast, total, thermo, hydro, wind, solar, bagasse] = hour_row
+
+        # "hour" is one-indexed
+        timestamp = get_timestamp(query_date=date, hour=hour - 1)
+        modes_extracted = [hydro, solar, wind, bagasse]
+
+        if total is None or None in modes_extracted:
+            continue
+
+        result.append(
+            zoneKey=zone_key,
+            datetime=timestamp,
+            production=ProductionMix(
+                hydro=hydro,
+                solar=solar,
+                wind=wind,
+                biomass=bagasse,
+                # NOTE: thermo includes gas + oil mixed, so we set these as unknown for now
+                # The modes here should match the ones we extract in the production payload
+                unknown=total - (hydro + solar + wind + bagasse),
+            ),
```

Comment:
Why are we not assigning thermo directly to unknown?
Or are there additional missing production?

Reply:
I think there might be additional unknowns we want to account for. This was already in the original parser.